Add Secret seed storage and key derivation spike

LLFourn opened the pull request:
resolves https://github.com/comit-network/comit-rs/issues/672

Note I only looked at lnd for inspiration. I couldn't find many other projects that did anything notable.

LLFourn committed:
Add Secret seed storage and key derivation spike

thomaseizinger commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+
+The wallet seed is stored encrypted with a password in the node's `wallet.db` (or at least I think that's where it's hiding).
+You have to unlock the wallet on lnd startup with the password via an RPC call (this is what the ln-cli does).
+There is a `--noseedbackup` option which id going to be depreciated which encrypts it with the default password so you don't have unlock the wallet.
```
```suggestion
There is a `--noseedbackup` option which is going to be depreciated which encrypts it with the default password so you don't have unlock the wallet.
```

thomaseizinger commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+### Use u8 rather than strings to identify key types
+
+There's nothing inherently wrong with using utf-8 encoded strings to a hash function to produce secret keys etc. e.g. `hash(seed || "secret")` (putting length extension attacks to the side).
+This is what we do now in rfc003's [SecretSource](https://github.com/comit-network/comit-rs/blob/master/application/comit_node/src/swap_protocols/rfc003/secret_source.rs).
```
:x: It would be good to use an absolute URL here (with a commit-hash instead of master) so that the link stays valid as we evolve the code.

thomaseizinger commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+
+..
+fn secp256k1_redeem(&self) -> KeyPair {
+    KeyPair::from_secret_key_slice(self.sha256_with_seed(&[SecretType::RedeemKey as u8]).as_ref())
```
:x: 

`as u8` depends on the order though I guess? We should probably just assign constant values to it to make it totally refactoring safe.

thomaseizinger reviewed (changes requested):
Really good work!

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+
+This section recommends how we should generate and store the secret seed by answering key questions.
+
+### Is there any way to recover anything if you have your secret seed but no database?
```
```suggestion
### Is there a way to recover anything if you have your secret seed but no database?
```

bonomat commented:
I'm about to start reviewing this PR, but in the meantime: have you seen this?
https://github.com/lightningnetwork/lnd/blob/master/docs/recovery.md

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+### Should it be encrypted?
+
+I think this should be optional.
+A COMIT node doesn't own the funds long term.
```
```suggestion
A COMIT node doesn't own any funds long term.
```

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+Having said that, giving an attacker access to your COMIT node will probably give them access to the HTTP API which they can use to redeem funds to their own address.
+This is probably a better attack vector than the above.
+
+note: We could prevent this attack by giving the node a way of generating the final redeem addresses without getting them from the api.
```
```suggestion
Note: We could prevent this attack by allowing the node to generate the final redeem addresses without getting them from the api.
```

LLFourn replied:
"giving" had the meaning I wanted to convey. In order for it to do this it would have to have a way to generate it given to it from the outside.

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+
+note: We could prevent this attack by giving the node a way of generating the final redeem addresses without getting them from the api.
+
+I think we should leave it unencrypted for now, make encryption an option later on and then perhaps make encryption the default as a last step.
```
```suggestion
For the sake of efficiency I propose to incrementally increase the security, i.e. (i) leave the secret unencrypted for now, (ii) make encryption an option later on and (iii) turn on encryption by default eventually.
```

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+### Where to store it?
+
+Unlike lnd, I don't think we should store secret seed in the database file.
+This allows users to backup the secret seed separately from the database (although they need to backup both).
```
👍I agree :)

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+Given that the two are linked, we should store a hash of the secret seed along with the swap id (in the same row or whatever).
+This allows the user to change the secret seed while not invalidating the entire database.
+
+By default I'd put it somewhere like:
```
I would rather not. If it is on a fixed location it's more vulnerable for attacks as the attacker would know where to look for the secret.

thomaseizinger replied:
The unix way to go seems to be a fixed location but restrict access to the file.
I like that better.

D4nte replied:
I don't think a fix location is an issue.

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
````diff
@@ -0,0 +1,195 @@
+~/.comit/secret_seed
+```
+
+But its location should be configurable with `secret_seed_file` in the `.toml` config.
````
👍That sounds like the best way to go.

D4nte replied:
I am happy to leave it to the user (ie, parano user choose a custom location, others choose the default one).

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+
+But its location should be configurable with `secret_seed_file` in the `.toml` config.
+
+The format of the file should be toml with a single entry `secret_seed` (so we can `encrypted_secret_seed` later).
```
What do you think of allowing to pass in an encrypted file itself instead?

thomaseizinger replied:
I am also in favor of just passing in an encrypted file that contains the seed in binary format. No need for a toml IMO.

LLFourn replied:
I'm not sure I understood the suggestions above. What about if it's not encrypted? Do you mean in the main config file you put `secret_seed_file = /path/to/file`  OR `encrypted_secret_seed_file = /path/to/encrypted_file`.

Although this would work, I think I'd prefer having it in a toml instead because then you can transition from encrypted to unencrypted without having to edit your main config file.

thomaseizinger replied:
Instead of having two toml entries, I'd rather design the seed-file itself in a way so that you can tell wether or not it is encrypted.

One way would be file-extension. A header in the file itself would also work.
In any case, that seems to be a much smoother way for the user to transition, since they don't have to change the config at all and there is always just one entry that refers to the `seed_file`.

LLFourn replied:
I'd prefer not to invent a new file format and just use something that exists.
Toml certainly might not be the best format!

D4nte replied:
Can someone explain what cons they see for a Toml file?

thomaseizinger replied:
> I'd prefer not to invent a new file format and just use something that exists.

I am pretty sure there are standard we can use for this.

> Can someone explain what cons they see for a Toml file?

Toml is a format for storing structured data. If all we have is a single entry, such a file format seems ill-fitted and overly complicated.

LLFourn replied:
@thomaseizinger The thing is we have at least two entries: `secret_seed` and `encrypted_secret_seed`. We could use [pem](https://docs.rs/pem/0.6.0/pem/) to label one or the other.

LLFourn replied:
I changed it to say the format should be text like the pem format.

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+I don't think it's worth encrypting the database yet until we have a motivation to do it.
+An attacker stealing the database would learn what protocol executions you're involved in so atm it's just a privacy issue.
+
+We *could* encrypt the database we the `secret_seed` seeing as they're separated.
```
Can you rephrase this sentence please (not sure if I understand it)?

thomaseizinger replied:
I think he wants to say:

```suggestion
We *could* encrypt the database with the `secret_seed` seeing as they're separated.
```

bonomat reviewed (commented):
I like it. Well done :)

thomaseizinger commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,195 @@
+
+### Where to store it?
+
+Unlike lnd, I don't think we should store secret seed in the database file.
```
The way this sentence is phrased confused me a bit :)

```suggestion
Unlike lnd, I think we should not store secret seed in the database file.
```

LLFourn committed:
Address comments #1

LLFourn committed:
Fix typo

LLFourn committed:
Add note about checksum

LLFourn commented:
@bonomat  No I hadn't. That page has the results of some nice lessons they've learned esp about how to backup. We could take their idea about having a restoration friendly version of the database that just has the minimal information needed to restore the protocol executions.

I've incorporated feedback (sometimes by re-writing the sentences that had suggestions). I also added a note about storing a checksum along with the seed.

D4nte commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,201 @@
+
+The format of the file should be toml with a single entry `secret_seed` (so we can `encrypted_secret_seed` later).
+
+We should make the permissions restrictive so only the user has read permission on it.
```
I assume you mean that if the unix permission are `og+r` then `comit_node` should reject the seed and fail to start? similar to ssh?

LLFourn replied:
Interesting idea. I wasn't thinking that actually. Just that when it's first created it should be created with restrictive permissions. It is much less clear in the case of comit what the correct permissions are for the file and I would just let the user decide but provide them with sane defaults.

D4nte commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,201 @@
+
+## Secret/NodeID Derivation
+
+I think the way we are doing it now is roughly the right way but I recommend modifying it in the following ways:
```
I expect the `NodeID` to be the `libp2p id`. What relation/synergy do you propose between both? How is the `libp2p id` set?

LLFourn replied:
see answer to @bonomat below.

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,201 @@
+
+This section recommends how we should generate and store the secret seed by answering key questions.
+
+### Can you recover anything if you have your secret seed but no database backup?
```
Anything in this case is too generic, I firstly wanted to ask you to explain what anything is, but then it became clear when I continued reading, hence, I think we can just remove it :)
```suggestion
### Can you recover if you have your secret seed but no database backup?
```

LLFourn replied:
I am asking: Is there anything you can recover from the secret seed by itself. 
Without it,  it implies that I am asking: Can you make a full recovery from the secret seed by itself.

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
````diff
@@ -0,0 +1,201 @@
+By default I'd put it somewhere like:
+
+```
+~/.comit/secret_seed
````
Our current configs are in: 
```
 ~/.config/comit_node/default.toml
```
should we put the `secret_seed` file in there as well?

LLFourn replied:
I'd say by default the `secret_seed` can go in the same place as the config file.

Hmm is `~/.config` a thing? It seems like it's more conventional to put things in `~/.comit`.

D4nte replied:
looks like `~/.config` is a JS thing as I have `yarn` and Electron stuff in it.
+1 for `~/.comit` across the board.

thomaseizinger replied:
`~/.config` is the default XDG_CONFIG_HOME directory and I guess the standard for most major linux distributions. https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html

thomaseizinger replied:
Most projects do whatever they want usually, hence home-folders are cluttered with `.my_random_project` folders. I am inclined to follow platform-friendly defaults and use something like https://github.com/soc/directories-rs to make it easily cross-platform.

On the other hand, that seems like a lost fight anyway, so if y'all vote for `.comit`, I am ok with that. Just wanted to raise some awareness :)

D4nte replied:
https://github.com/soc/directories-rs looks great

LLFourn replied:
I like the idea of using directories-rs, but putting crap in `/User/Name/Library` on mac is kind offensive. It gives us an answer for what to do on windows though which is nice.

D4nte replied:
I don't mind crap in `~/Library/` on Mac, isn't it the standard?

bonomat replied:
> On the other hand, that seems like a lost fight anyway, so if y'all vote for .comit, I am ok with that. Just wanted to raise some awareness :)

Don't give up that fast, you have me on your side: 
**+1 on `~/.config/comit_node`**

D4nte replied:
Let's aim for a `Win/Win` deal :)

LLFourn replied:
I removed mention of the path of secret_seed file and just said it should be next to the config file by default. Discussion can happen elsewhere about where to actually put the config file.

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,201 @@
+
+The implementation I recommend using is https://github.com/RustCrypto/hashes just because it's maintained by the rust crypto organisation.
+
+### NodeID
```
See comment from @D4nte above, I thought we already accepted the fact that the node ID is derived by libp2p?

LLFourn replied:
The node ID is derived from the key pair derived from the seed.

D4nte replied:
Are you saying there are **no** relations between the `NodeId` and the `libp2p id`?

thomaseizinger replied:
No. Libp2p creates an ID by deriving something from a keypair. We provide the keypair by deriving it from our seed.

D4nte replied:
I think I don't understand what is the point of this new `NodeID`

thomaseizinger replied:
The ID used by libp2p IS the node id!

D4nte replied:
> The ID used by libp2p IS the node id!

I expect this to be stated in the document and not something that takes 24 hours to understand/decrypt.

LLFourn replied:
I added the sentence: `This is the NodeID that will be used by libp2p.`.

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,201 @@
+
+The wallet seed is stored encrypted with a password in the node's `wallet.db` (or at least I think that's where it's hiding).
+You have to unlock the wallet on lnd startup with the password via an RPC call (this is what the ln-cli does).
+There is a (to be depreciated) `--noseedbackup` option  which encrypts it with the default password so you don't have unlock the wallet.
```
```suggestion
There is a (to be deprecated) `--noseedbackup` option  which encrypts it with the default password so you don't have unlock the wallet.
```

D4nte reviewed (changes requested):
See previous comments.

bonomat committed:
s/depreciated/deprecated/

Co-Authored-By: LLFourn <LLFourn@users.noreply.github.com>

LLFourn committed:
Clear up confusion about NodeID

LLFourn committed:
Add note about directories-rs

LLFourn committed:
Remove mention of ~/.comit/secret_seed

LLFourn committed:
Say the file format should be pem

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,201 @@
+
+Note: We could prevent this attack by giving the node a way of generating the final redeem addresses without getting them from the api.
+
+For the sake of expediency I propose we incrementally increase the security, i.e. (i) leave the secret unencrypted for now, (ii) make encryption an option later on and (iii) make encryption the default eventually.
```
👍

bonomat commented on 0003-secret-seed-storage-and-key-derivation.md:
```diff
@@ -0,0 +1,201 @@
+
+### How to generate it?
+
+If `secret_seed_file` isn't set, I think we should just generate it from `OsRng` to `~/.comit/secret_seed` if it doesn't exist.
```
We can rediscuss the location once we go for the implementation. Obviously, if we move to `~/.comit/`, then the config file should go there as well.

LLFourn replied:
I removed that mention of `~/.comit` replacing it with "its default location' :)

LLFourn committed:
Remove another mention of ~/.comit/secret_seed
